Refactor build process and switch to CircleCI

pcarranzav opened the pull request:
A rather big refactor of how the supervisor is built, to make it more usable and maintainable.
* Added a circle.yml file to build the supervisor on Circle
* Unified all the dockerfiles into a single multi-stage build
* Refactored the Makefile to reduce the amount of necessary variables to run a build, make use of the  multi-stage build, and document its usage
* Refactor jenkins_build.sh into a build.sh that is easier to use and is documented - and can also be used to build locally without pushing the artifacts.
* Refactor the dindctl script to take options instead of environment variables.
* Change the dind supervisor to use a local image instead of pulling an external one.
* Fix a problem in dind that breaks the host's ability to run terminals
* Add a deploy-to-resin.js that allows deploying to the Resin API as a supervisor_release. This is integrated into the CircleCI build if a master build is triggered with `DEPLOY_TO_RESIN=true`
* Add a script (automation/pr-to-meta-resin.sh) that automates creating PRs to meta-resin to bump the supervisor version. This one has to be run manually.

pcarranzav commented:
I think this is somewhat ready for review. The main things that I think are missing are:
* Add jobs for the missing architectures in circle.yml (for now I've only enabled i386 and amd64, but it's just a matter of copying the pattern)
* Give CircleCI credentials for `resin/` repos on dockerhub - @brownjohnf could you do this? I had tried with the DOCKER_PASSWORD you had set but I think it's outdated.
* Add `PUSH_IMAGES=true` to the calls to automation/build.sh on circle.yml, so that the builds actually push the resulting images - this will also make docker layer caching start working, so hopefully each build won't take three hours as they do now (since they constantly build the base image).

I might be missing something else though. #465, on the other hand, should add a decent amount of unit tests that we can integrate into the Circle builds too - ideally running the tests before going on to the full docker builds.

Page- commented:
@pcarranzav it would be huge if you could split this PR into separate commits - currently it's absolutely massive and very hard to review

pcarranzav commented:
@Page- done (as best as I could since the changes are a bit interdependent)

Page- commented on tools/dind/config/services/resin-supervisor-dind.service:
```diff
@@ -9,7 +9,7 @@ WorkingDirectory=/usr/src/app
 EnvironmentFile=/usr/src/app/config/localenv
 ExecStartPre=/bin/bash -c 'if [ "${PRELOADED_IMAGE}" == "true" ]; then /usr/bin/docker pull $(jq ".[0].imageId" ${APPS_PATH}); fi'
-ExecStartPre=/usr/bin/docker pull ${SUPERVISOR_IMAGE}
+ExecStartPre=/usr/bin/docker load -i /usr/src/supervisor-image.tar
```
Can you use `--input` please

pcarranzav replied:
Done

Page- commented on Dockerfile:
```diff
@@ -0,0 +1,207 @@
+RUN echo '#!/bin/sh\nexit 0' > /usr/bin/cross-build-end && chmod +x /usr/bin/cross-build-end
+
+# Build nodejs dependencies
+FROM $ARCH-node-base as node
```
Does this pick `$ARCH` up from the build args or is there another mechanism

pcarranzav replied:
Yup, it picks it from the build args

Page- commented on circle.yml:
```diff
@@ -46,14 +46,42 @@ jobs:
       DOCKER_USERNAME: travisciresin
```
Would it makes sense to abstract the standard env stuff to a default env section of the yaml?

Page- commented on Makefile:
```diff
@@ -120,7 +120,10 @@ ${DOCKERD_PROXY}:
 
-supervisor-dind: ${DOCKERD_PROXY}
+supervisor-tar:
+	docker save -o tools/dind/supervisor-image.tar $(SUPERVISOR_IMAGE)
```
Can you use `--output` please

pcarranzav replied:
Done

Page- commented on automation/Dockerfile:
```diff
@@ -4,7 +4,13 @@ RUN apk add --no-cache \
 WORKDIR /src
+COPY ./automation/package.json /src/automation/package.json
+RUN cd automation \
+	&& npm install
```
Can you add the `JOBS=MAX` and `npm cache clean` etc please

Page- commented on automation/deploy-to-resin.js:
```diff
@@ -0,0 +1,95 @@
+	method: 'GET'
+}, resinApi.passthrough))
+.then(function (config) {
+	let deviceTypes = config.deviceTypes;
```
```js
.then(({ deviceTypes }) => {

pcarranzav replied:
Done (no need for the {} now)

Page- commented on automation/deploy-to-resin.js:
```diff
@@ -0,0 +1,95 @@
+});
+
+resinApi._request(_.extend({
+	url: apiEndpoint + '/config',
```
If all you want is the device-types you should use `/config/device-types`

pcarranzav replied:
Nice, didn't know we had that one.  Done

Page- commented on automation/deploy-to-resin.js:
```diff
@@ -0,0 +1,95 @@
+			if (_.isEmpty(apiToken)) {
+				customOptions.apikey = apikey
+			}
+			console.log('Deploying ' + supervisor_version + ' for ' + device_type);
```
You can use
```es6
console.log(`Deploying ${supervisor_version} for ${device_type}`)
```
to have string interpolation

pcarranzav replied:
Done

Page- commented on automation/pr-to-meta-resin.sh:
```diff
@@ -0,0 +1,84 @@
+	exit 1
+fi
+
+USER=${USER:-`whoami`}
```
```bash
USER=${USER:-$(whoami)}
```
is preferable to `` ` `` because it allows for nesting and handles a bunch of other situations without surprises (it's not necessary in this case, but it's still a good habit to get into)

pcarranzav replied:
Done, and I'll read up about the difference

pcarranzav commented on Dockerfile:
```diff
@@ -0,0 +1,207 @@
+FROM resin/aarch64-node:6.5-slim as aarch64-node-base
+
+FROM resin/amd64-node:6.5-slim as amd64-node-base
+RUN echo '#!/bin/sh\nexit 0' > /usr/bin/cross-build-start && chmod +x /usr/bin/cross-build-start
```
That won't work as it would call the arm shell without emulator. We have to call cross-build-start in the `[ "command" ]` format so that it skips the shell.
My hope is that we'll have amd64 and i386 base images that already include a noop cross-build-start, but for now this is the only way. Good thing is this doesn't make it to the runtime image.

Page- replied:
Ok, that makes sense, can you combine the two `RUN`s into one instead then please

pcarranzav commented on automation/deploy-to-resin.js:
```diff
@@ -0,0 +1,95 @@
+// (if both are set, API_TOKEN is preferred).
+// The tag to deploy must be passed as $TAG.
+//
+var PlatformAPI = require('pinejs-client');
```
Good point, done

pcarranzav replied:
Done

pcarranzav commented on automation/deploy-to-resin.js:
```diff
@@ -0,0 +1,95 @@
+.then(function (config) {
+	let deviceTypes = config.deviceTypes;
+	// This is a critical step so we better do it serially
+	return Promise.mapSeries(deviceTypes, function (deviceType) {
```
Done

pcarranzav commented on automation/deploy-to-resin.js:
```diff
@@ -0,0 +1,95 @@
+			let device_type = deviceType.slug;
+			let customOptions = {}
+			if (_.isEmpty(apiToken)) {
+				customOptions.apikey = apikey
```
Done, I think

Page- commented on automation/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+WORKDIR /src
+COPY ./automation/package.json /src/automation/package.json
+RUN cd automation \
+	&& npm install
```
Lost from last review:
> Can you add the `JOBS=MAX` and `npm cache clean` etc please

Page- commented on automation/deploy-to-resin.js:
```diff
@@ -0,0 +1,94 @@
+	return Promise.mapSeries(deviceTypes, (deviceType) => {
+		if (deviceType.arch === arch) {
+			let image_name = `registry.resinstaging.io/resin/${arch}-supervisor`;
+			let supervisor_version = tag;
```
These should all be `const`s

Page- commented on automation/deploy-to-resin.js:
```diff
@@ -0,0 +1,94 @@
+				resource: 'supervisor_release',
+				body: {
+					image_name: image_name,
+					supervisor_version: supervisor_version,
```
This can be
```
body: {
	image_name,
	supervisor_version,
	device_type
}
```
although I think it would be better to just assign the object in-line rather than having js vars using a different naming style to our standard

Page- commented on automation/deploy-to-resin.js:
```diff
@@ -0,0 +1,94 @@
+					device_type: device_type,
+					is_public: true
+				},
+				customOptions: customOptions
```
This can be just
```
customOptions

pcarranzav commented:
All done @Page-

pcarranzav commented:
@Page- it took a few more changes to actually make it work - I removed the docker-in-docker builder and adjusted a few details to improve the build times. Still approve?
